crypto/x509/x509_vfy.c

```diff
@@ -494,23 +494,27 @@ static int check_chain(X509_STORE_CTX *ctx)
                          && ret != 1), ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
         }
-        if (num > 1) {
-            /* Check for presence of explicit elliptic curve parameters */
-            ret = check_curve(x);
-            CHECK_CB(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
-            CHECK_CB(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
-        }
+
         /*
-         * Do the following set of checks only if strict checking is requested
-         * and not for self-issued (including self-signed) EE (non-CA) certs
+         * Skip further checks for self-issued (including self-signed) EE certs
          * because RFC 5280 does not apply to them according RFC 6818 section 2.
          */
-        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-            && num > 1) { /*
-                           * this should imply
-                           * !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
-                           *          && (x->ex_flags & EXFLAG_SI) != 0)
-                           */
+        if (i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                && (x->ex_flags & EXFLAG_SI) != 0)
+            continue;
+
+        /* Reject unknown critical extensions due to RFC 5280 section 4.2 */
+        CHECK_CB((ctx->param->flags & X509_V_FLAG_IGNORE_CRITICAL) == 0
+                 && (x->ex_flags & EXFLAG_CRITICAL) != 0,
+                 ctx, x, i, X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION);
+
+        /* Check for presence of explicit elliptic curve parameters */
+        ret = check_curve(x);
+        CHECK_CB(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
+        CHECK_CB(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
```

Comment:
This check was previously only applied to non-EE certificates, i.e. those actually used in chain-integrity signature verification. It the application (not SSL, where this is not allowed) wants to allow use of explicit EC parameters, this is not the place to override that policy.

The only quibble is perhaps whether the check should be applied if we're checking "SS" signatures (even of EE certs)

Reply:
Actually, the check was previously done for _all_ certs in the chain if the chain had more than one element.
Note that this PR stops checking directly trusted EE certs already before reaching this point and on the other hand removes the `num > 1` guard that preceded these checks because I believe this guard has become pretty redundant.
If you want to have that it only applies to non-EE certificates, (which has not really the case before this PR), this check needs to be guarded differently.